### PR TITLE
Remove Pilot's sidecar

### LINKED ISSUE #6589: Remove Pilot's sidecar
Use the TLS/GRPC port configured by pilot for ADS instead of envoy proxy.
Once v1 is out, remove the sidecar and old ports, since it'll no longer be used.

### PR DESCRIPTION
**Pull request by ymesika**
Use the TLS/GRPC port configured by pilot for ADS instead of envoy proxy.
Once v1 is out, remove the sidecar and old ports, since it'll no longer be used.

### CODE REVIEW

**Comment by ymesika** [comment]
@rshriram No, I'm not sure.
Just examining what's affected and so far it indeed seems like tls grpc is the only issue.

**Comment by rshriram** [comment]
So I am all in favor of removing this if we add mTLS to the gRPC port in pilot. The pilot agent has code to do mTLS to pilot. So could use same code

**Comment by kyessenov** [comment]
@ymesika 

The issue at stake is putting mixer out of the loop for pilot telemetry. We have multiple telemetry backends (including GCP one) that depend on receiving data from mixer, and the pilot sidecar produces this data. Prometheus/grafana is only one option, and not necessarily the best one. Note that the tests in this repo may not even check for this any longer (lots of tests are still in semi-disabled state). Everything that the proxy does can be done in a library in pilot (duh), but we would need to align on the form of the telemetry, and how to consume it in mixer. You can start a document on direct API to mixer from pilot for this purpose, and probably how to integrate with citadel as well with a "library envoy".

**Comment by douglas-reid** [comment]
This seems like quite a large change. I'm sure that I've been out of the loop on the motivations, but are those documented somewhere (with discussion) ? The issue referenced appears to be just a reiteration of the PR description.

As @kyessenov relates, removing the proxy entirely drastically reduces any ability to monitor Pilot through existing Istio mechanisms. There obviously can be workarounds implemented, etc., but I'd like to have that thoroughly understood and committed to before jumping.

**Comment by ymesika** [comment]
The description has been taken from a list of work items in a pilot dev planing that @rshriram has put together and shared in a dedicated Slack channel. To be honest I might have misinterpreted the text in this PR's description and opened this PR today merely to investigate the effect of the change on the tests until getting further details.

This indeed seems like a major change which I'm not entirely sure what are the benefits beside simplification and smaller cpu/mem footprint. This led me to the thought that I didn't understand what was the work item about.

**Comment by douglas-reid** [comment]
@rshriram can we use an issue to track the work required here (with full discussion)? I was not part of the slack discussion -- and I imagine there are others like me that would like to be able to understand the whys, hows, and whens of these changes.

**Comment by rshriram** [comment]
We are not using the sidecar today. There is no support for extracting streaming gRPC metrics from mixer filter in Envoy. So all you see is just rudimentary metrics, thats not helpful to debug/monitor pilot in any manner. 

I do agree that we can stream metrics directly to mixer but thats purely additional work. Removing this proxy is not going to cause loss of pilot metrics as there are no metrics (for envoyv2) to start with. And we are already doing TLS in pilot. So, calling mixer over TLS should not be a big deal. There is no need to have any "citadel" library. Its just mTLS. 

While there was some passing mention of extracting streaming gRPC metrics via Envoy, its not implemented nor do I see any PRs doing this. A far more useful option here would be to simply import the metrics proto and make calls to mixer to stream the metrics from Pilot. You will get a lot more richer metrics than what you have today.

File: install/kubernetes/helm/istio/charts/pilot/templates/deployment.yaml [hunk 1]
```diff
@@ -67,49 +67,6 @@ spec:
           - name: istio-certs
             mountPath: /etc/certs
             readOnly: true
-        - name: istio-proxy
```
Comment (costinm) [inline install/kubernetes/helm/istio/charts/pilot/templates/deployment.yaml]: For upgrade from 0.8 (with alpha3) - you must also change the default port for grpc/mtls from 15012 to 15011.

**Comment by kyessenov** [comment]
@rshriram in what way does a sidecar in front of pilot prevent you from adding side-metrics to pilot?

**Comment by rshriram** [comment]
It doesn’t. But what’s the point of adding another moving part to the mix when the task can be accomplished with just pilot server ?

**Comment by andraxylia** [comment]
We are trying to get rid of the pilot-agent, not  add more functionality to it. SDS is coming and pilot agent will become less useful, so we can move to having a single process by container, whose lifecyle is controlled by kubelet. The fact pilot-agent controls envoy's lifecyle  is an anti-pattern.

Other than that, this exactly the kind of change we should not be making before cutting 1.0 in two weeks from now. There is no clear benefit, not a thorough design, and too much churn.

**Comment by costinm** [comment]
Well, pilot-agent is needed for other things and there are plans to use it for more ( for the future firewall, 
for secure tproxy, etc).

This is the kind of change to make for 1.0 - the envoy in front of pilot was a feature for v1 and we removed that. In v2 it does nothing of value - but adds overhead and complexity. 

Keeping some code that provides no value just because it may be do so in future is not a valid reason 
to block this PR. Once the 'direct' mixer API are available we can use them.

**Comment by kyessenov** [comment]
This is P1, and by far should not be high on anyone's list. By undoing the work we put in 0.8 for self-monitoring, you are just making the end goal that much harder.

**Comment by costinm** [comment]
There is no self-monitoring in v2 - all the interesting data is exported directly by pilot and is not visible in envoy. It is just complexity and wasted CPU/memory with no benefit.

**Comment by kyessenov** [comment]
Please update the conveniently disabled dashboard tests for istio components.

**Comment by douglas-reid** [comment]
I agree that streaming gRPC metrics are not going to provide a ton of value, especially in their current state. And I'm all in favor of developing a better app-local monitoring and finding an appropriate way to hook that up through Mixer.

However, I want to stress that the proxy was Reporting to Mixer for more than *just* metrics. Whatever solution is developed for replacing the metrics integration should include *full* Report integration.

My primary concern with this entire transition (from v1 with full Reporting to v2 with no reporting) is that it has happened without the replacement being ready. And that represents a loss in functionality -- a minor one in the grand scheme, perhaps, but one nonetheless.

A secondary concern is more about optics and "eating our own dogfood" but that is more philosophical and less tangible.

FWIW, I expect that in the 1.1 timeframe, we will have an improved, periodic reporting capability for streaming scenarios. When that is ready, Pilot will not be able to benefit from those improvements directly if the proxy is removed entirely.

In any case, I appreciate the discussion. I hope that we can quickly work towards a proper integration with Mixer in the near term. We should have at least two examples of writing a client for Mixer in golang at this point, so hopefully that work can be tackled without issue.

**Comment by rshriram** [comment]
>My primary concern with this entire transition (from v1 with full Reporting to v2 with no reporting) is that it has happened without the replacement being ready. And that represents a loss in functionality -- a minor one in the grand scheme, perhaps, but one nonetheless.

We lost most of metrics and reports when we moved to Envoy ads, even with the proxy in front. My point is that removing does not change this state in any way. To get the full set of metrics back, we can wait for 1.1+ and do nothing in pilot in the interim. 

Btw eating our own dog food does not imply we need to use a sidecar for applications that are capable of adhering to the mixer report protocol to send a set of metrics

**Comment by costinm** [comment]
@douglas-reid  I think we are on the same page and in strong agreement on the goals. 
The current plan for pilot is to use direct mixer APIs as soon as they are ready, and this PR is moving
into this direction by removing the non-functional metrics from envoy. Even if Envoy did have grpc
streaming - we believe direct mixer api will allow us to report richer metrics.

And it is going to 'eat our dogfood' as you say.

Right now mixer and pilot are only components using sidecar - if you believe more components
should eat the sidecar dogfood it would be great to add sidecar (and MTLS) to some of the 
others (prometheus, graphana, etc). Even Istio-CA, which I believe doesn't have one and is not
reporting anything ( and is not using streaming grpc ). It would be far more useful then trying to force
a particular solution (that no longer works) for pilot.

**Comment by costinm** [comment]
@Kuat: 
- citadel is a service, it exposes grpc cert renewal for mesh expansion
- we are not removing any functionality - in v2 the sidecar doesn't provide useful metrics,
just overhead. For 0.8 and v1 the sidecar provided metrics - for v2 it doesn't.

Yes, we did lose some metrics functionality when switching to v2 - but now v1 is gone 
and we'll gradually add back anything that is missing.

**Comment by costinm** [comment]
I would reiterate that prometheus, graphana, zipkin, etc are all suitable for use with envoy, they don't use streaming grpc and currently don't have mtls.

**Comment by costinm** [comment]
We can discuss this in the next networking WG meeting - probably ok to ship 1.0 without this, and merge it in master as we'll have more time to test and add any metrics that are identified by Kuat. 

Kuat: please include a list of metrics that envoy provides in 1.0 that would be lost, and we'll add them for 1.1 ( and please remove the hold).

**Comment by andraxylia** [comment]
I added this topic to the agenda for July 19.

**Comment by rshriram** [comment]
closed as duplicate

**Pull request closed without merging**